1. 你好，请教个问题，RNA-seq过程用Hisat2 align 以后，用cufflinks package 组装，其中Cuffdiff时一直报错Eorror:cannot open reference GTF file -L for reading.
2. `Traceback (most recent call last): .... ValueError: Error loading OTU table! If not in BIOM format use '-f' option`
3. `STAR --runThreadN 26 -- runMode genomeGenerate ...   Fatal INPUT FILE error, no valid exon lines in the GTF file: rRNA_tRNA.gtf   Solution: check the fomatting of the GTF file. Most likely cause is the difference in chromosome naming between GTF and FASTA file. `
4. `p <- ggplot(RAF,  aes(x=RAF, y=len)) + geom_boxplot()  Error in FUN(X[[i]], ...): object 'len' not found`
5. `大家好，我现在要做GO分析，但是现在我有的ID格式为locus_tag格式的，比如：Csa_3G120490请问用这种ID格式怎么做GO富集分析呢？如果这种格式不能做，该怎么转ID？`
6. `The reads contain too many k-mers to fit into available memory. You need approx. 420.229GB of free RAM to assemble your dataset`															   
7. `zsh conda: bad interpreter: /wtmp/user088/miniconda2/bin/python: no such file or directory`: python是否存在，特殊字符如^M
8. `export PATH=/wtmp/user088/miniconda2/bin/python`	中有什么错误？
9. 																	  

