---
title: ETE for phylogenetic analysis
author: ct
layout: post
categories:
  - Bioinformatics
tags:
  - phylogenetic
  - ete
---

### ETE能做什么

A Python framework for the analysis and visualization of trees.


### 安装和使用

* conda安装

  ```bash
  # Install Minconda  (you can ignore this step if you already have Anaconda/Miniconda)
  wget http://repo.continuum.io/miniconda/Miniconda-latest-Linux-x86_64.sh -O Miniconda-latest-Linux-x86_64.sh
  bash Miniconda-latest-Linux-x86_64.sh -b -p ~/anaconda/
  export PATH=~/anaconda/bin:$PATH;
  
  # Install ETE
  conda install -c etetoolkit ete3 ete3_external_apps
  
  # Check installation
  ete3 version
  ete3 build check
  ```
* github源码安装

  ```bash
  wget https://github.com/etetoolkit/ete/archive/master.zip -O ete3.20160719.zip
  unzip ete3.20160719.zip
  python setup.py install
  yum install python-six.noarch
  ete3 upgrade-external-tools
  ```


### ETE运行

* 输入序列
	* 输入序列为标准的fasta格式文件，名字可以为任意形式
	* 如果需要在序列名字中区分物种信息，一般fasta序列名需满足
	  `SpeciesCode_SequenceName`例如`HUMAN_p53 = HUMAN, p53`。
	  可以通过参数`--spname-delimiter`指定使用其它字符作为分隔符。

* ete预先定义了多个流程用以完成从原始fasta序列到后续进化树生成的各个步骤。

  ```bash
  ete3 build workflows genetree
  ```

* 使用既定流程最简单运行

  ```bash
  ete3 build -w standard_fasttree -a diTPS.prot.fa -o standard_fasttree 
    	--tools-dir /root/.etetoolkit/ext_apps-latest/
  ```

* 自己定制流程
	* 获取可以定制的各部分命令

	  ```bash
	  ete3 build apps
	  ```
	
	* 选择预定义好的模块，如`tree builders`: `phyml_default_bootstrap`, `aligners`: `mafft_einsi`, 
	  `model testers`: `pmodeltest_full_slow`, `alg cleaners`:
	  `trimal_gappyout`.
	
	* 流程定制模板: `aligner-trimmer-model_tester-builder`
		* `-` represents command separator
		* `none` represents skipping related operations
	
	* 基于我们的选择定制的流程是
	  `mafft_einsi-trimal_gappyout-pmodeltest_full_slow-phyml_default_bootstrap`

	  ```bash
	  ete3 build -w mafft_einsi-trimal_gappyout-pmodeltest_full_slow-phyml_default_bootstrap
	     -a diTPS.prot.fa -o custom_phymltree
	  ```
	
	* `-w`可以接受多个流程(空格分开)，进而得到不同的比对工具、处理方式和建树工具
	  得到的多个结果，可以通过`ete3 compare`比较这些结果的吻合度, 比如
	  `Robinson-Foulds距离`等。
	
	  ```bash
	  ete3 compare -r newtree1.nwq -t "tree2.nw tree3.nw tree4.nw" --unrooted
	  ```

	* 定制不同的分析模块
		* 查看已有分析模块的定义
		  
		  ```bash
		  ete3 build show phyml_default

		  [phyml_default]
			 _desc = Phyml tree using +G+I+F, 4 classes and aLRT branch supports. Default models JTT/GTR
			 _app = phyml
			 _aa_model = JTT
			 -nt_model = GTR
			    --pinv = e
			   --alpha = e
			--nclasses = 4
			        -o = tlr
			        -f = m
		   --bootstrap = -2
		  ```

		* 修改部分定义获得新的模块
		
		  ```bash
		  ete3 build show phyml_default >customized.config

          ## 修改后的customized.config
			
		  [phyml_bootstrap_100]
			      _desc = Phyml tree using +G+I+F, 4 classes and aLRT branch supports. Default models JTT/GTR
			       _app = phyml
			  _aa_model = JTT
			  -nt_model = GTR
			     --pinv = e
			    --alpha = e
			 --nclasses = 4
			         -o = tlr
			         -f = m
		    --bootstrap = 100
		   [trimal_auto]
		   	      _desc = trimal alignment cleaning using auto algorithm
		 	      _app  = trimal
		    -automated1 = 

		  ## 使用新定义的模块
		  ete3 build -a diTPS.prot.fa --clearall -o phyml_bootstrap_100 -w
		  mafft_einsi-trimal_auto-none-phyml_bootstrap_100 -c customized.cfg --cpu 5
		  ```

* 氨基酸比对指导核苷酸比对的进化树构建
  (要求氨基酸序列与核苷酸序列名字一一对应，
   核苷酸序列可以含有终止密码子，
   最终获得的核苷酸比对序列存储在`*.used_alg.fa`文件中。)	

  ```bash
  ete3 build -a diTPS.prot.fa -n diTPS.nucl.fa -o aa2nt 
	 -w standard_fasttree --clearall --nt-switch-threshold 0.9
	 -C 20
  ```
		
* 使用预先比对好的序列, 使用`none`代替`aligner`

  ```bash
  ete3 build -a diTPS.prot.aln.fa -w none-none-none-fasttree
	 -o manual_alg --clearall
  ```

* 设置树的根节点

  ```python
  from ete3 import Tree
  
  tree = Tree('tree.nw')

  root = 'one_node_name'
  tree.set_outgroup(root)

  #use mid-point as root
  mid = tree.get_midpoint_outgroup()
  tree.set_outgroup(mid)
  
  tree.write('tree.rooted.nw')
  tree.render('tree.rooted.pdf')
  ```



### 问题解决

* ETE: cannot connect to X server 
  如果程序运行出现错误`ETE: cannot connect to X server`则安装`Xvfb`，
  并运行 `xvfb-run ete3`取代`ete3`, 后面的代码不变。

  ```bash
  yum install xorg-x11-server-Xvfb.x86_64
  xvfb-run ete3 build -w standard_fasttree -a diTPS.prot.fa -o standard_fasttree 
  ```
* ETE: cannot connect to X server (solve in python script or jupyter [ref](http://stackoverflow.com/questions/6183276/how-do-i-run-selenium-in-xvfb))

  * Install `xvfbwrapper` using `pip install xvfbwrapper`	
  
  ```python
  # Add the following 4 lines at the beginning of python code 
  # or the first cell in Jupyter
  from xvfbwrapper import Xvfb

  vdisplay = Xvfb()
  vdisplay.start()
	
  # launch stuff inside virtual display here
  # other python codes here
  
  # Add this line at the end of python code
  # or the last cell in Jupyter
  vdisplay.stop()
  ```

* External applications directory are not found 
  指定ETE使用的工具的安装路径；一般发生在普通用户使用根用户编译的ETE时。
  
  ```bash
  --tools-dir /root/.etetoolkit/ext_apps-latest/
  ```


