---
title: Useful links
layout: page
comments: yes
---

### 准备GEO课程参考的教程网址

1. <https://warwick.ac.uk/fac/sci/moac/people/students/peter_cock/r/geo/>
2. <http://bioinformatics-core-shared-training.github.io/microarray-analysis/>
3. <https://github.com/bioinformatics-core-shared-training?page=2> 生信培训中心
4. <https://www.biostars.org/p/53870/>
5. <http://bioinformatics.knowledgeblog.org/2011/06/20/analysing-microarray-data-in-bioconductor/>
6. <http://bioinformatics.knowledgeblog.org/2011/06/21/volcano-plots-of-microarray-data/>
7. <https://wiki.bits.vib.be/index.php/Hands-on_Analysis_of_public_microarray_datasets>
8. <https://mp.weixin.qq.com/s?__biz=MzIyMzA2MTcwMg==&mid=2650732330&idx=1&sn=35751c91d8cbd62cf988679701f9da8e&chksm=f029a483c75e2d951591f06cfb923f8d92a5c30e2bb2e962f01a582aa159ad0a60520a208d9e&mpshare=1&scene=1&srcid=1201aeqQJfsBSaN71KyB8Dh9#rd>
9. <https://v.qq.com/x/page/w0159pm9ihy.html>
10. <https://www.researchgate.net/post/How_can_we_merge_different_Microarray_datasets_from_GEO_for_DEGs_Analysis>
11. <https://f1000research.com/articles/5-1384/v2>
12. 跨平台分析 <https://www.researchgate.net/post/What_software_is_best_for_cross-platform_microarray_results_comparison>
13. <https://bioinformatics.mdanderson.org/MicroarrayCourse/Lectures09/ma14b.pdf>
14. <https://wiki.bits.vib.be/index.php/Analyze_your_own_microarray_data_in_R/Bioconductor#Normalization_using_RMA>
15. <https://biochem.wisc.edu/sites/default/files/facilities/bcrf/tutorials/R/R_Bioconductor/CH25/from_cel_files_to_gene_list.html>
16. <http://math.usu.edu/jrstevens/stat5570/1.4.Preprocess_4up.pdf>
17. <https://bioconductor.org/help/course-materials/2009/SeattleApr09/AffyAtoZ/AffymetrixAtoZSlides.pdf>
18. <https://bioconductor.org/packages/devel/workflows/vignettes/maEndToEnd/inst/doc/MA-Workflow.html#9_rma_calibration_of_the_data>


### 准备TCGA课程教程参考

1. <https://bioconductor.org/packages/release/workflows/html/TCGAWorkflow.html>
2. <https://bioconductor.org/packages/release/bioc/html/TCGAbiolinks.html>
3. <https://f1000research.com/articles/5-1542/v2>
4. <http://www.sthda.com/english/articles/24-ggpubr-publication-ready-plots/77-facilitating-exploratory-data-visualization-application-to-tcga-genomic-data/> 很不错的可视化帖子
5. <https://bioinformaticsfmrp.github.io/Bioc2017.TCGAbiolinks.ELMER/index.html>
6. <https://github.com/greenelab/pancancer>
7. <https://github.com/inab/TCGA_visualizer>
8. <http://rtcga.github.io/RTCGA/articles/Visualizations.html>
9. <http://www.pathwaymapper.org/#>
10. <https://github.com/averissimo/tcga.data>
11. <http://www.liuzlab.org/TCGA2STAT/>
12. <http://r-addict.com/2016/11/21/Optimal-Cutpoint-maxstat.html>
13. <http://www.sthda.com/english/articles/24-ggpubr-publication-ready-plots/78-perfect-scatter-plots-with-correlation-and-marginal-histograms>
14. <https://blog.csdn.net/qq_29300341/article/details/74128979>
15. <https://blog.csdn.net/hzs106/article/details/12009915>
17. <http://paper.dxy.cn/article/511878> 一些概念可参考
16. <https://www.jianshu.com/p/54042789b0d5>
17. <https://blog.csdn.net/shmilyringpull/article/details/17529637>
18. <https://www.bioconductor.org/help/course-materials/2016/BioC2016/ConcurrentWorkshops4/Buros/applied-survival-model.html>
19. <http://bioinfostar.com/2017/12/07/Battlefield_TCGA-level3-RNAseq-data-01-zh_CN/>

### 待介绍的TCGA网站

1. <http://ualcan.path.uab.edu/index.html>
2. <https://cancer.sanger.ac.uk/cosmic/signatures>
2. TCGA文章合集

##### Most frequently visited sites
* [Tumblr](http://www.tumblr.com/dashboard)
* [Diggreader](http://digg.com/reader)
* [Quora](www.quora.com/)
* [Telegram](https://telegram.org/)
* [Khanacademy](https://www.khanacademy.org/profile/ChenTong/)

##### Blogs

* <https://liorpachter.wordpress.com/>
* Kalisto的作者 <http://sites.psu.edu/biomonika/page/2/>
* Nature系列有意思的工具 <http://blogs.nature.com/blog/tag/bioinformatics-2/>
* 基础教程 <https://davetang.org/muse/>

* [Kaggle, learn machine learning with examples](http://blog.kaggle.com/)
* [Blue Collar Bioinformatics,Brad Chapman，Harvard Public Health](http://bcbio.wordpress.com/)
* [simplystatistics](http://simplystatistics.org/)
* [gettinggeneticsdone](http://gettinggeneticsdone.blogspot.com/)
* [bioinformaticszen](http://www.bioinformaticszen.com/)
* [xiu shi jie](http://pgfe.umassmed.edu/ou)
* [Allen Downey](https://sites.google.com/site/allendowney/)

##### Tutorials

* [Data science](http://cm.dce.harvard.edu/2014/01/14328/publicationListing.shtml)
* [cs109 data science](http://cs109.org/)

##### Online courses

* [CS109](http://cs109.org/readings.php)
* [System Biology](http://ocw.mit.edu/courses/biology/7-342-systems-biology-stochastic-processes-and-biological-robustness-fall-2008/index.html)
* [machine learning](http://www.dcs.gla.ac.uk/~girolami/Machine_Learning_Module_2006/)

##### Visualization
* [Visualization](http://www-958.ibm.com/software/data/cognos/manyeyes/page/Visualization_Options.html)

##### Companies
* [seqWright](http://www.seqwright.com/)
*

##### Labs
* China
  * [brainnetome](http://www.brainnetome.org/en/)
  *

##### Bioliogy and disease

* Gene-disease
	
	* https://civic.genome.wustl.edu/ (curated, maybe best)
	* http://www.disgenet.org/web/DisGeNET/menu;jsessionid=8lztt2vozd7o8ey04ba9s0dk
	* http://snpedia.com/index.php/Category:Is_a_snp (larger)
	* http://www.cbioportal.org/ (https://www.biostars.org/p/179077/)

* Pipeline
	* https://github.com/ENCODE-DCC

* 国内博客

  * <http://www.bioinfo-scrounger.com/>  <gukai1212@163.com>

* Wiki
  * <https://wiki.bits.vib.be/index.php/Basic_bioinformatics_concepts,_databases_and_tools>
  * <https://www.biostars.org/p/142494/>
